Docs/security in gatsby (#22395)
gatsbyjs/gatsby: docs/docs/security-in-gatsby.md

---
title: Security in Gatsby
---

> By being a modern static site generator, Gatsby inherits several security principles that comes with this approach. However, even with this, your users could be exposed to malicious attacks. In this page, you will see what they are and how you could prevent them.

Reviewer: ```suggestion
By taking advantage of the benefits of static content, Gatsby inherits several security principles. However, even with this, your users can be exposed to malicious attacks. Read on for further information on potential attack vectors and how you could prevent them.
```

## Cross-Site Scripting (XSS)

> Cross-Site Scripting is a type of attack that injects a script or a unexpected link to another site into the client side of the application.

Reviewer: ```suggestion
Cross-Site Scripting is a type of attack that injects a script or an unexpected link to another site into the client side of the application.
```

> JSX elements automatically escape HTML tags by design. So, when you try to inject them as the example below:

Reviewer: ```suggestion
JSX elements automatically escape HTML tags by design. See the following example:
```

```js
// highlight-next-line
const username = `<script src='https://path/to/badness.js'></script>`

const User = () => <p> hello {username}</p>
```

> it will render `hello <script src='https://path/to/badness.js'></script>` as a string inside the p tag.

Reviewer: ```suggestion
When you try to inject the JSX element above it will render `hello <script src='https://path/to/badness.js'></script>` as a string inside the p tag.
```

> In other hand, there could be some fields in your application that you will need to render the inner HTML tags, such as a content field in a blog or a comment in a comments section, that are built in rich-text editors.

Reviewer: ```suggestion
On the other hand, fields in your application may need to render inner HTML tags, such as a content field in a blog or a comment in a comments section, that are built into rich-text editors.
```

> That's when you expose your application to XSS attacks, since the way to render these HTML tags is by using a HTML parser (e.g. [html-react-parser](https://github.com/remarkablemark/html-react-parser)) or using the `dangerouslySetInnerHTML` prop, as this example below:

Author: makes sense: https://english.stackexchange.com/questions/25742/is-it-supposed-to-be-a-html-or-an-html

> That's when you expose your application to XSS attacks, since the way to render these HTML tags is by using an HTML parser (e.g. [html-react-parser](https://github.com/remarkablemark/html-react-parser)) or using the `dangerouslySetInnerHTML` prop, as this example below:

Reviewer: ```suggestion
In order to render those HTML tags you need to use an HTML parser (e.g. [html-react-parser](https://github.com/remarkablemark/html-react-parser)) or the `dangerouslySetInnerHTML` prop, like so:
```

```js
const CommentRenderer = comment => (
  // highlight-next-line
  <p dangerouslySetInnerHTML={{ __html: comment }} />
) // dangerous indeed.
```

> 

Reviewer: ```suggestion
That's when you expose your application to XSS attacks. 
```

> **How can it be prevented?**

Reviewer: ```suggestion
### **How can you prevent cross-site scripting?**
```

> The most straightforward way to prevent the XSS attack is to sanitize the innerHTML string before dangerously setting it. Fortunately, there are npm packages that does the job, like [sanitize-html](https://www.npmjs.com/package/sanitize-html) and [DOMPurify](https://github.com/cure53/DOMPurify).

Reviewer: ```suggestion
The most straightforward way to prevent a XSS attack is to sanitize the innerHTML string before dangerously setting it. Fortunately, there are npm packages that can accomplish this; packages like [sanitize-html](https://www.npmjs.com/package/sanitize-html) and [DOMPurify](https://github.com/cure53/DOMPurify).
```

## Cross-Site Request Forgery (CSRF)

> A web application that use cookies could be attacked by the CSRF exploit, which deceives the browser to execute actions by the user's name without notice. By default, the browser "trusts" all the activity made validating the user's identity and therefore sending the associated cookies in every request.

Author: In these past few days, I'm exploring more in the topic in [CS253 - Web Security classes from Stanford](https://web.stanford.edu/class/cs253/). I'll rewrite this section, making it simple, doing some quotes and linking these resources.

> For example, assume that the comments done in your blog are sent in a form like this below:

Reviewer: ```suggestion
For example, assume that the comments in your blog are sent in a form similar to this one:
```

```html
<form action="http://mywebsite.com/blog/addcoment" method="POST">
  <input type="text" name="comment" />
  <input type="submit" />
</form>
```

> A malicious website could inspect your site and copy it to theirs. If the user are logged in, it still works because when the form is sent, the associated cookies goes with it and the action is made. Even worse, the form could be sent when the page loads with informations you don't control:

Reviewer: ```suggestion
A malicious website could inspect your site and copy it to theirs. If the user is logged in, it still works because when the form is sent, the associated cookies goes with it and the action is made. Even worse, the form could be sent when the page loads with information you don't control:
```
Not sure I understand the middle sentence. Is the goal to explain that this is true for unauthenticated AND authenticated users?

Author: Sorry, it's confusing. It only works if you are authenticated to the original website. I'm going to rewrite this section.

```js
// highlight-next-line
<body onload="document.csrf.submit()">
<!-- ... -->
<form action="http://mywebsite.com/blog/addcoment" method="POST" name="csrf" >
  // highlight-next-line
  <input type="hidden" name="comment" value="Hey visit http://maliciouswebsite.com, it's pretty nice" />
  <input type="submit" />
</form>
```

> **How can it be prevented?**

Reviewer: ```suggestion
### **How can you prevent cross-site request forgery?**
```

> **How can it be prevented?**

Author: Ok, sorry about that! I think that is good we are learning things while doing this. I'm gonna rewrite this section.

> **How can it be prevented?**

Reviewer 2: no need to be sorry. you're doing great work, just trying to give helpful feedback. thanks and keep it up. I've already learned a thing or two.

> **How can it be prevented?**

Author: You are!

> ### CSRF Tokens

Reviewer: ```suggestion
#### CSRF Tokens
```

> In a page you want to protect, your server will provide an encrypted, hard to guess **token**, which is tied to the user's session and thus will be required to be sent back in the POST form:

Reviewer: ```suggestion
If you want to protect a page your server will provide an encrypted, hard to guess **token**. That token is tied to a user's session and is must be included in every POST request. See the following example:
```

```js
<form action="http://mywebsite.com/blog/addcoment" method="POST">
  <input type="text" name="comment" />
  // highlight-next-line
  <input type="hidden" name="token" value="SU9J3tMoT1w5q6uJ1VMXaaf9UXzLvyNd" />
  <input type="submit" />
</form>
```

> So when the form is sent, the server will compare them and block the action if they are not the same. Note that usually the malicious website won't have access to this CSRF token because of [HTTP Access Control](https://developer.mozilla.org/pt-BR/docs/Web/HTTP/Headers/Access-Control-Allow-Origin).

Reviewer: ```suggestion
When the form is sent, the server will compare the token received with the stored token and block the action if they are not the same. This works because malicious websites don't have access to the CSRF token due to [HTTP Access Control](https://developer.mozilla.org/pt-BR/docs/Web/HTTP/Headers/Access-Control-Allow-Origin).
```

> ### Same-Site Cookies Directive

Reviewer: ```suggestion
#### Same-Site Cookies Directive
```

If you need to create cookies in your application, make sure to protect them by adding the `SameSite` directive:

`Set-Cookie: example=1; SameSite=Strict`

It allows the server to make sure that the cookies are not being sent by a **cross-site** domain request.

> Check out [MDN Docs](https://developer.mozilla.org/pt-BR/docs/Web/HTTP/Headers/Set-Cookie) regarding how a cookie could be configured. You can checkout current browser support over on the [Can I Use page](https://caniuse.com/#feat=same-site-cookie-attribute).

Reviewer: ```suggestion
Check out [MDN Docs](https://developer.mozilla.org/pt-BR/docs/Web/HTTP/Headers/Set-Cookie) for more information on configuring a cookie. You will also want to note current browser support which is available on the [Can I Use page](https://caniuse.com/#feat=same-site-cookie-attribute).
```

## Third-party Scripts

> Some third-party scripts like Google Tag Manager give you the ability to [add arbitrary JavaScript](https://support.google.com/tagmanager/answer/6107167) to your site. This helps integrate third-party tools but can be misused to inject malicious code. To avoid this, be sure to [control access ](https://support.google.com/tagmanager/answer/6107011) to these services.

Reviewer: ```suggestion
Some third-party scripts like Google Tag Manager give you the ability to [add arbitrary JavaScript](https://support.google.com/tagmanager/answer/6107167) to your site. This helps integrate third-party tools but can be misused to inject malicious code. To avoid this, be sure to [control access](https://support.google.com/tagmanager/answer/6107011) to these services.
```

## Check Your Dependencies

> In your Gatsby project, you are going to have a lot of dependencies (which has their own dependencies as well) in your `node_modules/`. Therefore, it is important to check if any of them has a security issue.

Reviewer: ```suggestion
In your Gatsby project, you are going to have some dependencies that get stored in `node_modules/`. Therefore, it is important to check if any of them, or their dependencies, have security vulnerabilities.
```

> **Using `npm`**

Reviewer: ```suggestion
### Using npm
```

> In npm, you can use the `npm audit` command to check your dependencies. It is available since v6. Check their [docs](https://docs.npmjs.com/cli/audit) for more options.

Reviewer: ```suggestion
In npm, you can use the `npm audit` command to check your dependencies. This command is available in all npm versions after `6.0.0`. Check [npm docs](https://docs.npmjs.com/cli/audit) for more options.
```

Author: ```suggestion
In npm, you can use the `npm audit` command to check your dependencies. This command is available in all npm versions greater than `6.0.0`. Check [npm docs](https://docs.npmjs.com/cli/audit) for more options.
```
What do you think?

> **Using `yarn`**

Reviewer: ```suggestion
### Using yarn
```

> Similar to `npm`, you can use the `yarn audit` command. It is available since v1.12.0, not available in v2. Check their [docs](https://classic.yarnpkg.com/en/docs/cli/audit/) for more options.

Reviewer: ```suggestion
Similar to npm, you can use the `yarn audit` command. It is available starting with version `1.12.0` though it is not yet available in version 2. Check the [yarn docs](https://classic.yarnpkg.com/en/docs/cli/audit/) for more options.
```

## Key Security

> Gatsby provides the capability of [fetching data from various APIs](/docs/content-and-data/) and those APIs often require a key to access them. These keys should be stored in the build environment using [Environment Variables](/docs/environment-variables/) as shown below when fetching data from GitHub with an Authentication Header:

Reviewer: ```suggestion
Gatsby allows you to [fetch data from various APIs](/docs/content-and-data/) and those APIs often require a key to access them. These keys should be stored in your build environment using [Environment Variables](/docs/environment-variables/). See the following example for fetching data from GitHub with an Authentication Header:
```

> Gatsby provides the capability of [fetching data from anywhere](/docs/content-and-data/) and by that you should know that most of this data requires a key to access them. These keys should be stored in the server by using [Environment Variables](/docs/environment-variables/) as this example below, fetching data from GitHub with an Authentication Header:

Reviewer 2: ```suggestion
Gatsby provides the capability of [fetching data from various APIs](/docs/content-and-data/) and those APIs often require a key to access them. These keys should be stored in the build environment using [Environment Variables](/docs/environment-variables/) as shown below when fetching data from GitHub with an Authentication Header:
```

Reviewer 2: Also a link to the appropriate section of https://www.gatsbyjs.org/docs/environment-variables/ would be good.

```js
    {
      resolve: "gatsby-source-graphql",
      options: {
        typeName: "GitHub",
        fieldName: "github",
        url: "https://api.github.com/graphql",
        headers: {
          // highlight-next-line
          Authorization: `Bearer ${process.env.GITHUB_TOKEN}`,
        },
      }
    }
```

> Note: Whether you need to authenticate someone in your application, Gatsby has an [Authentication Tutorial](/tutorial/authentication-tutorial) which helps doing this job in a secure way.

Reviewer: ```suggestion
> Note that Gatsby has an [Authentication Tutorial](/tutorial/authentication-tutorial) if you need assistance with setting up authentication on your Gatsby site in a secure way.
```

> Note: Whether you need to authenticate someone in your application, Gatsby has an [Authentication Tutorial](/tutorial/authentication-tutorial) which helps doing this job in a secure way.

Reviewer 2: This covers build variables well but not client side keys. Not sure if the note is intended to cover it cause the language is a bit confusing.

Author: Yes, this note does not cover this part. What do you think it should be here regarding client-side keys? encryption? known methods, e.g. JWT?

Reviewer 2: I think this doc should be a starting point. so defining the issue and maybe an example of how to solve then (again) links out to more thorough resources (like Gatsby's docs on keys and env variables).

## Content Security Policy (CSP)

Content Security Policy is a security layer added in web applications to detect and prevent attacks, e.g. the XSS attack mentioned above.

To add it to your Gatsby website, add [gatsby-plugin-csp](/packages/gatsby-plugin-csp/) to your `gatsby-config.js` with the desired configuration. Note that
currently there is a [incompatibility issue](https://github.com/gatsbyjs/gatsby/issues/10890) between [gatsby-plugin-csp](/packages/gatsby-plugin-csp/) and [gatsby-image](/packages/gatsby-image) and other plugins that generate hashes in inline styles.

> Note: Not all browsers support CSP, check [can-i-use](https://caniuse.com/#feat=mdn-http_headers_csp_content-security-policy) for more information.

Reviewer: ```suggestion
> Note that not all browsers support CSP, check [can-i-use](https://caniuse.com/#feat=mdn-http_headers_csp_content-security-policy) for more information.
```

const User = () => <p> hello {username}</p>
```

When you try to inject the JSX element above it will render `hello <script src='https://path/to/badness.js'></script>` as a string inside the p tag.

On the other hand, fields in your application may need to render inner HTML tags, such as a content field in a blog or a comment in a comments section, that are built into rich-text editors.

In order to render those HTML tags you need to use an HTML parser (e.g. [html-react-parser](https://github.com/remarkablemark/html-react-parser)) or the `dangerouslySetInnerHTML` prop, like so:

```js
const CommentRenderer = comment => (
  // highlight-next-line
  <p dangerouslySetInnerHTML={{ __html: comment }} />
) // dangerous indeed.
```

That's when you expose your application to XSS attacks.

### **How can you prevent cross-site scripting?**

The most straightforward way to prevent a XSS attack is to sanitize the innerHTML string before dangerously setting it. Fortunately, there are npm packages that can accomplish this; packages like [sanitize-html](https://www.npmjs.com/package/sanitize-html) and [DOMPurify](https://github.com/cure53/DOMPurify).

## Cross-Site Request Forgery (CSRF)

> Cross-Site request forgery is a type of exploit that which deceives the browser to execute unauthorized actions. By default, in any request made, the browser automatically appends any stored cookies of the destination domain. Combining this with a crafted request, an malicious website can read and write data without the user's action or knowledge.

Author: No, actually the browser sends all the cookies together with the request. Is "append" the word for that?

```html
<img src="http://mywebsite.com/blog/addcomment?comment=unwanted%20comment" />
```

#### CSRF Tokens

If you want to protect a page your server will provide an encrypted, hard to guess **token**. That token is tied to a user's session and is must be included in every POST request. See the following example:

```js
<form action="http://mywebsite.com/blog/addcoment" method="POST">
  <input type="text" name="comment" />
  // highlight-next-line
  <input type="hidden" name="token" value="SU9J3tMoT1w5q6uJ1VMXaaf9UXzLvyNd" />
  <input type="submit" />
</form>
```

When the form is sent, the server will compare the token received with the stored token and block the action if they are not the same. Make sure that malicious websites don't have access to the CSRF token by using [HTTP Access Control](https://developer.mozilla.org/pt-BR/docs/Web/HTTP/Headers/Access-Control-Allow-Origin).

#### Same-Site Cookies Attribute

This cookie attribute is targeted to prevent CSRF attacks. If you need to create a cookie in your application, make sure to protect them by this attribute, that could be of `Strict` or `Lax` type:

> `Set-Cookie: example=1; SameSite=Strict`

Author: I don't think so :/ This is a HTTP header.

> `Set-Cookie: example=1; SameSite=Strict`

Reviewer: Perhaps using `text` then?

> `Set-Cookie: example=1; SameSite=Strict`

Author: like this?
```
Set-Cookie: example=1; SameSite=Strict
```

> `Set-Cookie: example=1; SameSite=Strict`

Reviewer: Something like 

```text
Set-Cookie: example=1; SameSite=Strict
```

> `Set-Cookie: example=1; SameSite=Strict`

Author: Sure! just added a code block for this.

> `Set-Cookie: example=1; SameSite=Strict`

Author: Is the text after ``` required or just ``` is fine?

> `Set-Cookie: example=1; SameSite=Strict`

Reviewer: The text is what we add for code blocks that aren't in a specific language. So I'd prefer that we add it.

> `Set-Cookie: example=1; SameSite=Strict`

Author: Added!

> It allows the server to make sure that the cookies are not being sent by a **cross-site** domain request.

Author: ```suggestion
Using the `SameSite` attribute allows the server to make sure that the cookies are not being sent by a **cross-site** domain request.
```

Check out [MDN Docs](https://developer.mozilla.org/pt-BR/docs/Web/HTTP/Headers/Set-Cookie) for more information on configuring a cookie. You will also want to note current browser support which is available on the [Can I Use page](https://caniuse.com/#feat=same-site-cookie-attribute).

Quoting [OWASP](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#samesite-cookie-attribute):

> > It is important to note that this attribute should be implemented as an additional layer defense in depth concept. This attribute protects the user through the browsers supporting it, and it contains as well 2 ways to bypass it as mentioned in the following section. This attribute should not replace having a CSRF Token. Instead, it should co-exist with that token in order to protect the user in a more robust way. Also remember that any Cross-Site Scripting (XSS) can be used to defeat all CSRF mitigation techniques!

Reviewer: I'm wondering if we need this full quote? Is the goal to explain that you need to do this in addition to the CSRF token? If so, perhaps it should be mentioned in that section?

Author: The goal is to explain that the `SameSite` cookies attribute is not a replacement for CSRF Token (and vice-versa) and they can work to together as security layers in a website. Also, if there is Cross-Site Scripting (XSS) attack, both of them are useless to a CSRF attack. Is this the best place to put this information? Should it stay as a quote or I should write a new paragraph?

Reviewer: I think what you just wrote here is clearer. So perhaps a paragraph and then a link to that information if they want to read more?

Author: Yes, makes sense. I replaced this quote with a paragraph, adding the link as well.

Reviewer: The new text is no longer a quote, correct?

Author: correct!

### Using npm

In npm, you can use the `npm audit` command to check your dependencies. This command is available in all npm versions greater than `6.0.0`. Check [npm docs](https://docs.npmjs.com/cli/audit) for more options.

### Using yarn

Similar to npm, you can use the `yarn audit` command. It is available starting with version `1.12.0` though it is not yet available in version 2. Check the [yarn docs](https://classic.yarnpkg.com/en/docs/cli/audit/) for more options.

## Key Security

Gatsby allows you to [fetch data from various APIs](/docs/content-and-data/) and those APIs often require a key to access them. These keys should be stored in your build environment using [Environment Variables](/docs/environment-variables/). See the following example for fetching data from GitHub with an Authorization Header:

```js
    {
      resolve: "gatsby-source-graphql",
      options: {
        typeName: "GitHub",
        fieldName: "github",
        url: "https://api.github.com/graphql",
        headers: {
          // highlight-next-line
          Authorization: `Bearer ${process.env.GITHUB_TOKEN}`,
        },
      }
    }
```

### Storing keys in client-side

> Sometimes in your Gatsby website, you will need display sensitive data or handle authenticated routes (e.g. a page that shows a user's orders in your ecommerce). Gatsby has an [Authentication Tutorial](/tutorial/authentication-tutorial) if you need assistance with setting up authentication flow. Use cookies to store the credentials in client-side, preferably with `SameSite` attribute listed above. Check out [MDN Docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies) to further explanation about these attributes and how to configure them.

Reviewer: ```suggestion
Sometimes in your Gatsby website, you will need display sensitive data or handle authenticated routes (e.g. a page that shows a user's orders in your ecommerce). Gatsby has an [Authentication Tutorial](/tutorial/authentication-tutorial) if you need assistance with setting up authentication flow. Use cookies to store the credentials client-side, preferably with the `SameSite` attribute listed above. Check out [MDN Docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies) to further understand these attributes and how to configure them.
```

## Content Security Policy (CSP)

Content Security Policy is a security layer added in web applications to detect and prevent attacks, e.g. the XSS attack mentioned above.

To add it to your Gatsby website, add [gatsby-plugin-csp](/packages/gatsby-plugin-csp/) to your `gatsby-config.js` with the desired configuration. Note that

> currently there is a [incompatibility issue](https://github.com/gatsbyjs/gatsby/issues/10890) between [gatsby-plugin-csp](/packages/gatsby-plugin-csp/) and [gatsby-image](/packages/gatsby-image) and other plugins that generate hashes in inline styles.

Reviewer: ```suggestion
currently there is a [compatibility issue](https://github.com/gatsbyjs/gatsby/issues/10890) between [gatsby-plugin-csp](/packages/gatsby-plugin-csp/) and other plugins that generate hashes in inline styles, including [gatsby-image](/packages/gatsby-image).
```

---
title: Security in Gatsby
---

By being a modern static site generator, Gatsby inherits several security principles that comes with this approach. However, even with this, your users could be exposed to malicious attacks. In this page, you will se what they are and how you could prevent them.

## Cross-Site Scripting (XSS)

Cross-Site Scripting is a type of attack that injects a script or a unexpected link to another site into the client side of the application.

JSX elements automatically escapes HTML tags by design. So, when you try to inject them as the example below:

```js
// highlight-next-line
const username = `<script src='https://path/to/badness.js'></script>`

const User = () => <p> hello {username}</p>
```

it will render `hello <script src='https://path/to/badness.js'></script>` as a string inside the p tag.

In other hand, there could be some fields in your application that you will need to render the inner HTML tags, such as a content field in a blog or a comment in a comments section, that are built in rich-text editors.

> That's when you expose your application to XSS attacks, since the only way to render these HTML tags in React is by using the `dangerouslySetInnerHTML` prop:

Reviewer 2: This isn't entirely true. there are several other packages that do this. [html-to-react](https://www.npmjs.com/package/html-to-react) and [html-react-parser](https://www.npmjs.com/package/html-react-parser) are the two I've used. I think these both sanitize on the client side, but not on SSR (but don't quote me on that...). They offer a lot of additional features too but those are not inherently related to security. I'd include this information some how so folks no it's an option...I'll leave it to you exactly how this should be done.

Author: You are right. There are alternatives to `dangerouslySetInnerHTML` but they don't sanitize HTML. see https://github.com/remarkablemark/html-react-parser/issues/94 and https://github.com/aknuds1/html-to-react/issues/79.

I reformulate the phrase to mention the html-to-react package, what do you think?

Reviewer 2: I'd at least mention them.  I'm fairly sure html-react-parser doesn't do script tags on the client side (I ran into this on a project). one could also use the custom parsing functionality in these to sanitize html

Author: I mentioned html-react-parser! Please check again and see that if it is good.

```

**How can it be prevented?**

The most straightforward way to prevent the XSS attack is to sanitize the innerHTML string before dangerously setting it. Fortunately, there are npm packages that does the job, like [sanitize-html](https://www.npmjs.com/package/sanitize-html) and [DOMPurify](https://github.com/cure53/DOMPurify).

## Cross-Site Request Forgery (CSRF)

A web application that use cookies could be attacked by the CSRF exploit, which deceives a your user to execute actions by its name. By default, the browser "trusts" all the activity made validating the user's identity and therefore sending the associated cookies in every request.

For example, assume that your users received an email with a malicious link:

```html
<a href="http://maliciouswebsite.xyz/">Get your coupon here</a>
```

Then, when the website opens, it could render this tag:

```html
<img src="http://yourecommerce.com/profile" />
```

which, instead of rendering an image, it sends the request with the authentication cookies and opens a security breach in their personal information.

**How can it be prevented?**

This could be prevented in three major ways:

### Additional Session Token

> In addition to cookie validation, another approach could be added in the flow, such as an [OAuth2](https://oauth.net/2/).

Reviewer 2: seems a little brief, I'm not understanding how this helps. maybe expand a little here.

Author: Ops, I misunderstand the CSRF Token solution to this problem. Removing this part.

### Same-Site Cookies

Configure the created cookie in a way that it could be only sent by the same domain. [Most moderns browsers support it with a backward compatibility](https://caniuse.com/#feat=same-site-cookie-attribute).

### Anti-CSRF Tokens

> 

Reviewer 2: Maybe needs an example. not totally understanding how/why this works.

Author: Yes, totally. I add a example about how to configure it

In addition to a form, a hidden input could be added, holding the value of a encrypted token generated by the serve when page loads. When the form is sent, the server can block the action if they are not the same.

## Third-party Scripts

> Some third-party scripts like Google Tag Manager gives the possibility to add arbitrary JavaScript code on top of them. Be sure to control the access of these services.

Reviewer 2: This definitely needs some explanation. How do I control access, to the third-party scripts, or to GTM?

Author: Absolutely. I expanded with 2 docs links to Google

## Check Your Dependencies

> In your Gatsby project, you are going to have a lot of dependencies (which has their own dependencies as well) in your `node_modules/`. Therefore, it is important to check if any of them has a security issue. NPM make this possible since v6 by executing the `npm audit` in the command line.

Reviewer 2: I'd add the yarn equivalent. and may be worth noting the yarn v1 and v2 version if they differ.

Author: yarn v2 does not support `audit` yet. I changed the writing to talk about yarn as well.

## Content Security Policy (CSP)

Content Security Policy is a security layer added in web applications to detect and prevent attacks, e.g. the XSS attack mentioned above.

> To add it to your Gatsby website, add the [plugin](https://www.gatsbyjs.org/packages/gatsby-plugin-csp/) to your `gatsby-config.js` with the desired configuration. Note that

Reviewer 2: ```suggestion
To add it to your Gatsby website, add [gatsby-plugin-csp](https://www.gatsbyjs.org/packages/gatsby-plugin-csp/) to your `gatsby-config.js` with the desired configuration. Note that
```

> currently there is a [incompatibility issue](https://github.com/gatsbyjs/gatsby/issues/10890) with [gatsby-image](https://www.gatsbyjs.org/packages/gatsby-image) and other plugins.

Reviewer 2: Expound on  "other plugins"...I understand it's unreasonable to list every incompatible plugin, but please define _why_ or _what_ makes a plugin incompatible.

Author: Sure thing! I wrote the reason after that, which is more detailed in the issue.

> Also note that [it works in most of browsers](https://caniuse.com/#feat=mdn-http_headers_csp_content-security-policy).

Reviewer 2: ```suggestion
Note:  Not all browsers support CSP, check [can-i-use](https://caniuse.com/#feat=mdn-http_headers_csp_content-security-policy) for more information.
```
